developer/src/kmldmlc/src/keyman/kmx/kmx-plus-builder.ts

```diff
@@ -0,0 +1,319 @@
+    };
+  }
+
+  private build_strs(): BUILDER_STRS {
+    return {
+      ident: constants.hex_section_id(constants.section.strs),
+      size: 0,  // finalized later
+      _offset: 0,
+      count: 0,  // finalized later
+      reserved: 0,
+      items: [], // finalized later
+    };
+  }
+
+  private build_meta(): BUILDER_META {
+    return {
+      ident: constants.hex_section_id(constants.section.meta),
+      size: constants.length_meta,
+      _offset: 0,
+      name: this.alloc_string(this.file.kmxplus.meta.name),
+      author: this.alloc_string(this.file.kmxplus.meta.author),
+      conform: this.alloc_string(this.file.kmxplus.meta.conform),
+      layout: this.alloc_string(this.file.kmxplus.meta.layout),
+      normalization: this.alloc_string(this.file.kmxplus.meta.normalization),
+      indicator: this.alloc_string(this.file.kmxplus.meta.indicator),
+      settings: this.file.kmxplus.meta.settings ?? 0,
+    };
+  }
+
+  private build_loca(): BUILDER_LOCA {
+    let loca: BUILDER_LOCA = {
+      ident: constants.hex_section_id(constants.section.loca),
+      size: constants.length_loca + constants.length_loca_item * this.file.kmxplus.loca.locales.length,
+      _offset: 0,
+      count: this.file.kmxplus.loca.locales.length,
+      reserved: 0,
+      items: []
+    };
+
+    for(let item of this.file.kmxplus.loca.locales) {
+      loca.items.push(this.alloc_string(item));
+    }
+
+    return loca;
+  }
+
+  private build_keys(): BUILDER_KEYS {
+    if(!this.file.kmxplus.keys.keys.length) {
+      return null;
+    }
+
+    let keys: BUILDER_KEYS = {
+      ident: constants.hex_section_id(constants.section.keys),
+      size: constants.length_keys + constants.length_keys_item * this.file.kmxplus.keys.keys.length,
+      _offset: 0,
+      count: this.file.kmxplus.keys.keys.length,
+      reserved: 0,
+      items: []
+    };
+
+    for(let item of this.file.kmxplus.keys.keys) {
+      keys.items.push({
+        vkey: item.vkey,
+        mod: item.mod,
+        // todo: support 'extend'
+        to: this.alloc_string(item.to),
+        flags: 0 // todo: support extend
+      });
+    }
+
+    return keys;
+  }
+
+  private build_vkey(): BUILDER_VKEY {
+    if(!this.file.kmxplus.vkey.vkeys.length) {
+      return null;
+    }
+
+    let vkey: BUILDER_VKEY = {
+      ident: constants.hex_section_id(constants.section.vkey),
+      size: constants.length_vkey + constants.length_vkey_item * this.file.kmxplus.vkey.vkeys.length,
+      _offset: 0,
+      reserved: 0,
+      count: this.file.kmxplus.vkey.vkeys.length,
+      items: []
+    };
+
+    for(let item of this.file.kmxplus.vkey.vkeys) {
+      vkey.items.push({
+        vkey: item.vkey,
+        target: item.target
+      });
+    }
+
+    return vkey;
+  }
+
+  private finalize_strs() {
+    this.sect_strs.count = this.sect_strs.items.length;
+    let offset = constants.length_strs + constants.length_strs_item * this.sect_strs.count;
+    // TODO: consider padding
```

Comment:
maybe

```suggestion
    offset += (0x10-(offset%0x10))%0x10;
```

but probably should be a function, and should make sure to zero

Reply:
I'll do padding in a follow-up PR, but will consider something like this.